test/integration/test_basic.py

```diff
@@ -14,7 +14,4 @@ def juju():
 def test_deploy(juju):
     juju.deploy('snappass-test')
 
-    def is_active(status):
-        return status.applications['snappass-test'].application_status.current == 'active'
-
-    juju.wait_status(is_active)
+    juju.wait(lambda status: status.is_app_active('snappass-test'))
```

Comment:
In Matrix you said that you felt that the helper-returning-a-function option wouldn't work with combining or composing, so it has to be like this.

I agree this makes something like this pretty readable:

```python
    juju.wait(lambda status: (status.is_app_active('snappass-test') or status.is_app_waiting('snappass-test'))
```

But you can compose the other way too, like this (or other ways, e.g. decorator style wrapping):

```python
    juju.wait(lambda status: (is_app_active('snappass-test')(status) or is_app_waiting('snappass-test')(status)))
```

Or this sort of thing:

```python
    juju.wait(lambda status: any(f(status) for f in is_app_active('snappass-test'), is_app_waiting('snappass-test')))
```

Reply:
I find that second example hard to understand, and the third one quite difficult to get my head around. It's something about the extra level of indirection. I had to go over it a few times to convince myself it would work. Whereas I find the first one easy to understand (though no need for `(` after `lambda status: `).

Would be interested to see if others agree, or if it's just me.

Reply:
I quite like the `Status.is_app_active` style helpers. Have you considered something like:
```py
@dataclass
class JujuState:
    applications: list[Application]
```
Which would let you put helpers on the  application objects themselves
```py
juju.wait(lambda state: app.is_active() for app in state.applications)
```
If we're committed to the `dict` structure, the helpers could go on `ApplicationStatus`
```py
juju.wait(lambda status: app.is_active() for name, app in state.applications.items())
```

Reply:
Keeping the `dict` would allow
```py
status.is_app_active('snappass-test')
```
written as
```py
status.applications['snappass-test'].is_active()
```

Reply:
Interesting idea. Did you mean to have `all(...)` in those lambdas, or are you thinking `wait` would do an implicit `all()` if the callback returned a generator? I think I'd vote against that and keeping it less magical, as it's unclear whether it'd be `all()` or `any()`.

In any case, the list is an interesting idea. Pros: makes certain operations like this simpler. Cons: no longer 1:1 with the structure returned by Juju, and also you can't look up by name directly.

Reply:
Keeping the dict -- yeah, exactly.

Ah, interesting idea to have the `is_active` on the `ApplicationStatus` and `UnitStatus` object. Then I guess we could shorten `applications` to `apps` (as I've done elsewhere) and get:

```python
status.apps['snappass-test'].is_active()

# which is very comparable to:

status.is_app_active('snappass-test')
```

I'll give it some more thought.

Reply:
Sweet. If there's any value in having a `list` instead of a `dict` (like iteration over the rich object instead of string name by default), you could still allow lookup by name with something like `status.get_app('app-name')`

Reply:
I did mean to have an explicit `all`/`any` btw, not suggesting magic

Reply:
Thanks. I've added `ApplicationStatus.is_<status>` helper properties (same for `UnitStatus`) and removed the `is_app_active()` methods on the Status object. I've also renamed "applications" to "apps" everywhere for brevity (it's a very clear abbreviation and widely used). It's one character longer but I think it's now cleaner -- thanks! I've made them properties as they no longer need an argument.

```
# previously:
status.is_app_active('snappass')

# now:
status.apps['snappass'].is_active
```